language/libra-tools/writeset-transaction-generator/src/lib.rs

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use libra_types::{
+    account_address::AccountAddress,
+    transaction::{Script, Transaction},
+};
+use libra_types::{account_config::libra_root_address, transaction::WriteSetPayload};
+use std::io::Write;
+use stdlib::compile_script;
+use tempfile::NamedTempFile;
+
+fn compile_admin_script(input: &str) -> Result<Script> {
+    let mut temp_file = NamedTempFile::new()?;
+    temp_file.write_all(input.as_bytes())?;
+    let cur_path = temp_file.path().to_str().unwrap().to_owned();
+    Ok(Script::new(compile_script(cur_path), vec![], vec![]))
+}
+
+pub fn encode_remove_validators_transaction(validators: Vec<AccountAddress>) -> Transaction {
+    assert!(validators.len() > 0, "Unexpected validator set length");
+    let mut script_body = "".to_string();
+    for address in validators {
+        script_body.push_str(
+            format!(
+                "LibraSystem::remove_validator(libra_root, 0x{:?});\n",
+                address
+            )
+            .as_str(),
+        );
+    }
+
+    let script = compile_admin_script(
+        format!(
+            "
+    script {{
+        use 0x1::LibraSystem;
+        fun main(libra_root: &signer) {{
+            {}
+        }}
+    }}",
+            script_body
+        )
+        .as_str(),
+    )
+    .expect("Cannot compile script");
+
+    Transaction::GenesisTransaction(WriteSetPayload::Script {
```

Comment:
there's no authentication for GenesisTransaction, is there somewhere we stop processing it if it's sent through consensus instead of applying manually?

Reply:
oh I think consensus only accepts `SignedTransaction`, halt_network seems more useful as signed writeset? or in general do we want the tooling to be used by AOS too?

Reply:
That is exactly right! The removing validator functionablity would be only doable in the GenesisTransaction setting if we lose enough validators to form a quorem. The `halt_network`, as you suggested, can be sent from AOS as well and should be how it would be used in general.